### ISSUE TITLE
match_explorer query: some problems with IndexReader.numDocs() while calculating idf

### ISSUE THREAD

**Issue by ndkmath1** [o19s/elasticsearch-learning-to-rank#375]
Now the plugin is using this way to calculate idf: https://github.com/o19s/elasticsearch-learning-to-rank/blob/main/src/main/java/com/o19s/es/explore/ExplorerQuery.java#L111
`idf_stats.add(sim.idf(tStats.docFreq(), searcher.getIndexReader().numDocs()));`

There're three problems

1. TermStatistics.docFreq() counts deleted document but IndexReader.numDocs() doesn't  (link: https://lucene.apache.org/core/8_9_0/core/org/apache/lucene/search/TermStatistics.html.  "Values may include statistics on deleted documents that have not yet been merged away.")
In Lucene, to calculate idf, "...CollectionStatistics.docCount() is used instead of IndexReader#numDocs()" (link: https://lucene.apache.org/core/8_9_0/core/org/apache/lucene/search/similarities/TFIDFSimilarity.html#idfExplain-org.apache.lucene.search.CollectionStatistics-org.apache.lucene.search.TermStatistics-)

2. numDocs() doesn't depend on search_type (query_then_fetch and dfs_query_then_fetch), it always counts on shard level for both search_type but docFreq() depends on search_type
- if search_type is dfs_query_then_fetch, docFreq will count on index level
- if search_type is query_then_fetch, docFreq will count on shard level

Therefore, number of documents should also depend on search_type.

3. if mappings has nested field, numDocs() will count hidden documents. Imagine that my shard has 1 document and this document has a nested field for arrays of 100 objects. Then numDocs() is equal 101. (link: https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html)

These three problems will be solved if we do as follows (I can help with a pull request if it makes sense)
`idf_stats.add(sim.idf(tStats.docFreq(), searcher.collectionStatistics(term.field()).docCount()));`

Do the problems and the way to solve it make sense? Let me know if I need to clarify anything.

**Comment by KhanhNguyen6** [comment]
> 2\. docFreq() depends on search_type
> 
> * if search_type is dfs_query_then_fetch, docFreq will count on shard level
> * if search_type is query_then_fetch, docFreq will count on index level

I had a small typo in (2). It should be
- if search_type is dfs_query_then_fetch, docFreq will count on index level
- if search_type is query_then_fetch, docFreq will count on shard level
I can provide a small example about this problem if needed.

**Comment by ndkmath1** [comment]
> 2\. docFreq() depends on search_type
> 
> * if search_type is dfs_query_then_fetch, docFreq will count on shard level
> * if search_type is query_then_fetch, docFreq will count on index level

I had a small typo in (2). It should be
- if search_type is dfs_query_then_fetch, docFreq will count on index level
- if search_type is query_then_fetch, docFreq will count on shard level

Therefore, number of documents should also be depend on search_type. I can provide a small example about this problem if needed.

**Comment by worleydl** [comment]
Great catches!

I think this all makes sense would it ever be useful to have the existing "broken" statistic under another name? I could see your number 3 as being useful in some cases but you're right in that it's not the intended behavior here.

PR's are always welcome!

**Comment by ndkmath1** [comment]
Thanks for your response.

> I think this all makes sense would it ever be useful to have the existing "broken" statistic under another name? 

Could you please clarify this sentence? Maybe you want to see an example that idf behavior has broken?

>  I could see your number 3 as being useful in some cases but you're right in that it's not the intended behavior here.

Could you tell more about some cases that count hidden will useful?

**Comment by nathancday** [comment]
Hi @ndkmath1, thanks for opening this issue too. Just wanted to chime in and clarifying the solve you suggest.

1. I'm confused Lucene's `CollectionStatistics.docCount`, *may* include [delete flagged](https://lucene.apache.org/core/8_1_1/core/org/apache/lucene/search/CollectionStatistics.html) docs. I don't think this is a problem, its just the cost of doing business w/ Lucene.

2. Are you proposing to expose the underlying `search_type` in Elasticsearch's [search APIs](https://www.elastic.co/guide/en/elasticsearch/reference/7.13/search-template.html#search-template-api-query-params)? Using `dfs_query_then_fetch` will add in an extra phase to the query which may slow things down. 

3. Nested fields are separate docs. I'm not sure there is a good way to get around this other than by considering the field(s) in context. Maybe I am missing your use case.

I can't speak for Dan, but PR's are welcome and would help to show what you are proposing.

**Comment by worleydl** [comment]
> > I think this all makes sense would it ever be useful to have the existing "broken" statistic under another name?
> 
> Could you please clarify this sentence? Maybe you want to see an example that idf behavior has broken?
I just have a concern that some users may want to continue using the broken behavior (even if it's not typical for idf).  I wouldn't worry about it if you do make a PR.  It was more of a question/concern as we review this issue.

I don't have any specific examples off hand.  I only mention it as it could throw a kink into existing models trained off datasets with lots of nested docs.

**Comment by ndkmath1** [comment]
Hi @nathancday, @worleydl 

I did a [PR](https://github.com/o19s/elasticsearch-learning-to-rank/pull/378) about workaround for the problems.

Below's something what I want to discuss with @nathancday 

> 1. I'm confused Lucene's CollectionStatistics.docCount, may include delete flagged docs. I don't think this is a problem, its just the cost of doing business w/ Lucene.


I mean, [the current idf of the plugin](https://github.com/o19s/elasticsearch-learning-to-rank/blob/main/src/main/java/com/o19s/es/explore/ExplorerQuery.java#L111) 

`                    idf_stats.add(sim.idf(tStats.docFreq(), searcher.getIndexReader().numDocs()));
`

- ` tStats.docFreq()` counts deleted documents
- `searcher.getIndexReader().numDocs()` doesn't count deleted document

 We should make both in the same direction. I mean numDocs() should also count deleted documents. To do it, we can use `CollectionStatistics.docCount() instead of IndexReader.numDocs()`. We can do it with my [PR](https://github.com/o19s/elasticsearch-learning-to-rank/pull/378). 
 
 

>  2. Are you proposing to expose the underlying search_type in Elasticsearch's search APIs? Using dfs_query_then_fetch will add in an extra phase to the query which may slow things down.

Yes, Elasticsearch supports `dfs_query_then_fetch`. Should we make `match_explorer` support it? Sometimes we want to have precise value and need to use `dfs_query_then_fetch`. Now if I do this request
 ```
POST my_index/_search?search_type=dfs_query_then_fetch
 {
    "query": {
        "match_explorer": {
            "type": "sum_classic_idf",
            "query": {
                "match": {
                    "title": "when breath becomes air"
                }
            }
        }
 }
```
I see the plugin is using docFreq over index, docCount over shard. We just need to make docCount in the same direction (over index). My [PR](https://github.com/o19s/elasticsearch-learning-to-rank/pull/378) can do it.

>  3. Nested fields are separate docs. I'm not sure there is a good way to get around this other than by considering the field(s) in context. Maybe I am missing your use case.

Yes, we can ignore nested docs by considering the field. Same as my [PR](https://github.com/o19s/elasticsearch-learning-to-rank/pull/378)

**Comment by nathancday** [comment]
Ok that makes perfect sense, thanks. I like your solution.

**Comment by worleydl** [comment]
Fixed and released in 1.5.6

**Comment by ndkmath1** [comment]
Thank you.
